dll/shellext/netplwiz/SHDisconnectNetDrives.c

```diff
@@ -0,0 +1,229 @@
+/*
+ * COPYRIGHT:       See COPYING in the top level directory
+ * PROJECT:         ReactOS Disconnect Network Drive
+ * FILE:            dll/shellext/netplwiz/SHDisconnectNetDrives.c
+ * PURPOSE:         Disconnects mapped network drives.
+ * PROGRAMMERS:     Jared Smudde
+ */
+
+#include "netplwiz.h"
+#include <wine/debug.h>
+
+WINE_DEFAULT_DEBUG_CHANNEL(shell);
+
+HINSTANCE hInstance;
+
+VOID InitializeListView(HWND hDlg)
+{
+    HWND hListView = GetDlgItem(hDlg, IDC_CONNECTEDDRIVELIST);
+    LV_COLUMN column;
+
+    if(hListView == NULL)
+        return;
+
+    column.mask = LVCF_WIDTH | LVCF_TEXT;
+    column.pszText = L"Drive Letter";
+    column.cx = 75;
+    ListView_InsertColumn(hListView, 0, &column);
+
+    column.mask = LVCF_WIDTH | LVCF_TEXT;
+    column.pszText = L"Network Path";
+    column.cx = 150;
+    ListView_InsertColumn(hListView, 1, &column);
+}
+
+VOID EnumerateConnectedDrives(HWND hDlg)
+{
+    LV_ITEM item;
+    HIMAGELIST hImgList;
+    HICON hIconDrive = NULL;
+    HMODULE hShell32;
+    HWND hListView = GetDlgItem(hDlg, IDC_CONNECTEDDRIVELIST);
+
+    DWORD dRet;
+    HANDLE hEnum;
+    LPNETRESOURCE lpRes;
+    DWORD dAllocSize = 0x1000;
+    DWORD dCount;
+    LPNETRESOURCE lpCur;
+
+    /* List View Icons */
+    hShell32 = GetModuleHandleW(L"shell32.dll");
+
+    if (hShell32 == NULL)
+        return;
+
+    hIconDrive = LoadImageW(hShell32, MAKEINTRESOURCEW(10), IMAGE_ICON,
+                            GetSystemMetrics(SM_CXSMICON),
+                            GetSystemMetrics(SM_CYSMICON), 0);
+
+    if (hIconDrive == NULL)
+        return;
+
+    hImgList = ImageList_Create(16, 16, ILC_COLOR32 | ILC_MASK, 1, 1);
+
+    if (hImgList == NULL)
+    {
+        DestroyIcon(hIconDrive);
+        return;
+    }
+
+    ImageList_AddIcon(hImgList, hIconDrive);
+    DestroyIcon(hIconDrive);
+    ListView_SetImageList(hListView, hImgList, LVSIL_SMALL);
+
+    dRet = WNetOpenEnum(RESOURCE_CONNECTED, RESOURCETYPE_DISK, 0, NULL, &hEnum);
+    if (dRet != NO_ERROR)
+    {
+        ERR("WNetOpenEnum failed with error: 0x%08x\n", dRet);
+        return;
+    }
+
+    lpRes = HeapAlloc(GetProcessHeap(), 0, dAllocSize);
+
+    if (lpRes == NULL)
+    {
+        ERR("Could not allocate memory.\n", dRet);
+        WNetCloseEnum(hEnum);
+        return;
+    }
+
+    do
+    {
+        DWORD dSize;
+        dCount = -1;
+        memset(lpRes, 0, dAllocSize);
+
+        dRet = WNetEnumResource(hEnum, &dCount, lpRes, &dSize);
+
+        if (dRet == WN_MORE_DATA)
+        {
+            PVOID pBuff = HeapReAlloc(GetProcessHeap(), 0, lpRes, dSize);
+
+            if (pBuff == NULL)
+            {
+                ERR("Could not reallocate memory.\n", dRet);
+                break;
+            }
+
+            lpRes = pBuff;
+            dAllocSize = dSize;
+
+            continue;
```

Comment:
This looks rather suspicious and might explain why it doesn't work on W2K3.

Can you try something like: https://git.reactos.org/?p=reactos.git;a=blob;f=base/applications/network/net/cmdUse.c;hb=HEAD ?

I'd be interested to know whether it work better on W2K3.

Reply:
If I remember correctly, this is where I got my code from.

Reply:
Alright. I tried the code from net use and it still didn't work under 2003.

Reply:
Well, it's at least better than current code ;-).

Reply:
The only downside is that the listview items don't display properly :)